modules/core/shared/src/main/scala/Trace.scala

```diff
@@ -79,9 +79,12 @@ object Trace {
             child <- parent.span(name, options)
           } yield new (IO ~> IO) {
             def apply[A](fa: IO[A]): IO[A] =
-              local
-                .set(child)
-                .bracket(_ => fa.onError(child.attachError(_)))(_ => local.set(parent))
+              local.get.flatMap { old =>
+                local
+                  .set(child)
+                  .bracket(_ => fa.onError(child.attachError(_)))(_ => local.set(old))
+              }
```

Comment:
The `parent` span for this `Resource` is not necessarily the span that was installed on the `IO` at point of transformation.